aws_xray_sdk/ext/pymongo/patch.py

```diff
@@ -0,0 +1,36 @@
+from pymongo import monitoring
+from aws_xray_sdk.core import xray_recorder
+
+
+class XrayCommandListener(monitoring.CommandListener):
+    """
+    A listener that traces all pymongo db commands to AWS Xray.
+    Creates a subsegment for each mongo db conmmand.
+
+    name: 'MongoDB-127.0.0.1:27017'
+    annotations: command_name:str, database_name: str, succeed: bool
+    """
+
+    def started(self, event):
+        host, port = event.connection_id
+        subsegment = xray_recorder.begin_subsegment(
+            f'MongoDB-{host}:{port}', 'remote')
+        subsegment.put_annotation('command_name', event.command_name)
+        subsegment.put_annotation('database_name', event.database_name)
+
+    def succeeded(self, event):
+        xray_recorder.end_subsegment()
+
+    def failed(self, event):
+        subsegment = xray_recorder.current_subsegment()
+        subsegment.add_fault_flag()
+        subsegment.put_metadata('failure', event.failure)
```

Comment:
Can the failure event type be pickled all the time? The `pymongodb` API reference doc doesn't have very clear information based on what I looked at here: http://api.mongodb.com/python/current/api/pymongo/monitoring.html#pymongo.monitoring.CommandFailedEvent. If not then the failure event serialization might fail randomly at the runtime. See more information here: https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled. You could consider manually serialize it before add it to the metadata.

Reply:
I digged into PyMongo's implementation. Yes it is unspecified in the documentation, but I suspect it may not change for no reason.
The `failure` field is assigned when PyMongo receives a remote error, in format of BSON. PyMongo decodes it into a python `dict` and wrap into an `OperationError`.

Anyway, the `failure` property contains a `dict`-typed BSON Object, and is pickle-able.

I did an experiment:

```py
try:
    a = client.a.command('')
except pymongo.errors.OperationFailure as err:
    d = err.details
    print(d, type(d), d.keys())
```
Result:
```
{'ok': 0.0, 'errmsg': 'no such command: \'\', bad cmd: \'{ : 1, $readPreference: { mode: "secondaryPreferred" }, $db: "a" }\'', 'code': 59, 'codeName': 'CommandNotFound'} <class 'dict'> dict_keys(['ok', 'errmsg', 'code', 'codeName'])
````

Reply:
I'm fine with this since official doc is not clear about the failure model.

Reply:
By the way, in the documents there may be pymongo types `bson.son.SON` and `bson.objectid.ObjectId`. They *can* be pickled, but how will X-Ray backend handle them?